src/backend/linux_raw/fs/raw_dir.rs

```diff
@@ -0,0 +1,201 @@
+    #[inline]
+    #[doc(alias = "inode_number")]
+    pub fn ino(&self) -> u64 {
+        self.inode_number
+    }
+
+    /// Returns the seek cookie to the next directory entry.
+    #[inline]
+    #[doc(alias = "off")]
+    pub fn next_entry_cookie(&self) -> i64 {
+        self.next_entry_cookie
+    }
+}
+
+impl<'buf, Fd: AsFd> Iterator for RawDir<'buf, Fd> {
+    type Item = io::Result<RawDirEntry<'buf>>;
+
+    #[allow(unsafe_code)]
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.offset < self.initialized {
+                let dirent_ptr = self.buf[self.offset..].as_ptr();
+                // SAFETY:
+                // - This data is initialized by the check above.
+                //   - Assumption: the kernel will not give us partial structs.
+                // - Assumption: the kernel uses proper alignment.
+                let dirent = unsafe { &*dirent_ptr.cast::<linux_dirent64>() };
+
+                self.offset += usize::from(dirent.d_reclen);
+
+                return Some(Ok(RawDirEntry {
+                    file_type: dirent.d_type,
+                    inode_number: dirent.d_ino,
+                    next_entry_cookie: dirent.d_off,
+                    file_name: {
+                        let name_start = dirent.d_name.as_ptr().cast::<u8>();
+                        let mut name_len = {
+                            // On 32-bit platforms, the kernel continues to use 8 byte alignment,
+                            // so cannot rely on size_of or align_of and must instead hardcode
+                            // that knowledge in.
+                            const ALIGNMENT: usize = 8;
+                            const DIRENT64_PACKED_SIZE: usize = 19;
+
+                            // Find the last non-padding byte of the file name so we can
+                            // start searching for NUL bytes. If we started searching
+                            // directly from the back, we would run into garbage left over
+                            // from previous iterations.
+                            usize::from(dirent.d_reclen)
+                                .saturating_sub(DIRENT64_PACKED_SIZE + ALIGNMENT)
+                        };
+
+                        // SAFETY:
+                        // - We start searching from within the dirent AND before the padding bytes.
+                        // - The kernel guarantees a NUL terminated name.
+                        while unsafe { *name_start.add(name_len) } != 0 {
+                            name_len += 1;
+                        }
+                        // Add 1 for the NUL byte
+                        name_len += 1;
+
+                        // SAFETY: We found the NUL byte and length above.
+                        unsafe {
+                            let file_name = CStr::from_bytes_with_nul_unchecked(
+                                slice::from_raw_parts(name_start, name_len),
+                            );
```

Comment:
Instead of searching for the NUL above and then doing `from_bytes_with_nul_unchecked` here, could this use `CStr::from_ptr`?

Reply:
~We could, but that would make me sad (I actually have a debug_assert that uses from_ptr). The current implementation creates the CStr in constant time whereas using from_ptr will scale with the length of the file name.~

Son of a biscuit, premature optimization strikes again. from_raw is annoyingly faster than my constant-time thing because it uses avx2 instructions. Done. :)

I tested it with file names that were ~30 bytes long which I'd actually consider to be longer than average.